### DISCUSSION TITLE
Summary_plot with LSTM and GRU models

### DISCUSSION THREAD

**Discussion by St-Savin** [shap/shap#2251] [Q&A]
I did modify the format of the shap values given by DeepExplainer so I could use Summary_plot. I was wondering if it was correct to do so.  The model is for a multivariate temporal (from t0 to th) model and the way I programmed it I can isolate the summary_plot just for the variables at t0 (or any timesplice between t0 and th) or take all the variables (t0 to th).  To me it seems correct since variables like temperature are less important at t=0 than t=1 which is logical for my model. I also know that only selecting the summaryplot for a timeslice is not very informative about the model but we can compare all the variable from the same timeslice.

**Comment by jainraj** [comment]
Two points: 

1. (if using PyTorch) I would suggest against using DeepExplainer for LSTM and GRU - DeepExplainer uses DeepLIFT which I think isn't supported for these models yet. [unsure about TF] 
2. I agree with you that looking at all variables at all timesteps would make more sense than single timesteps.

**Reply by St-Savin** [reply]
1. I can't help you with that point, I am barely functional in TF and I keep forgetting everything that I learn since I mainly work with tabular data and XGboost does the job quite well. 
2. I will try to make a very simple model to better analyze the issue and isolate it from any conflict with TF, pytorch, DeepExplainer. That way,  we will be able to see if there is any issues with a timestep or timeslice. By the way, putting all the time slices together would make you miss important time effects like delays. The analysis of the simple model will help to determine what timescales are important.

**Comment by LeoZWT** [comment]
I also agree. Can anyone provide good advice on how to calculate the SHAP of variables with different time steps over time

**Reply by St-Savin** [reply]
What do you mean different time steps ? Like random time steps [(x1,y1,t)] [(x2,y2,1)(x3,y3,4)(x4,y4,7)(x5,y5,9)(x6,y6,13) ] or time steps that follow a sequence (+1 +3 +6) [(x2,y2,1)(x3,y3,3)(x4,y4,6)(x5,y5,7)(x6,y6,9)(x6,y6,12) ] ?

**Reply by DaniBrkr** [reply]
@LeoZWT you might want to try the old API of SHAP, I managed to extract the SHAP values from a mixed model (LSTM + FC), contrary to the new way of doing it

x consists of [x_time_series, x_static] with dimensions [batch_size x time steps x features_ts] and [batch_size x features_static],
getting back the values with the same structure

```
 explainer = shap.DeepExplainer(
        model=model,
        data=x_train_explain,
    )

shap_values = explainer.shap_values(x_test_explain)
```

**Reply by St-Savin** [reply]
I did some[ tests](https://github.com/St-Savin/SHAP_times-series) on times series models and their SHAP value, and there is a problem with model with 2d input (like LSTM). The sampling frequency is set by the fastest variables and the lenght of the window is set by the slowest variable. Therefore, it created a big input that will drown the inpact of the slowest variables and will require more ressources to train. If all the variables have the same dynamics, it won't be an issue.  As for your specific case with different time step over time, I would resample your data to have the same time-step. Then, you can use whatever model you need, even XGBoost.
